garden_ai/app/notebook.py

```diff
@@ -37,6 +39,57 @@
 
 
 class DockerClientSession:
+    def __init__(self, verbose=False) -> None:
+        self.verbose = verbose
+
+    def __enter__(self) -> docker.DockerClient:
+        try:
+            return get_docker_client()
+        except DockerStartFailure as e:
+            # We're most likely to see this error raised from get_docker_client.
+            self.handle_docker_start_failure(e)
+        except docker.errors.BuildError as e:
+            self.handle_docker_build_failure(e)
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if exc_type is DockerStartFailure:
+            # If the user's Docker daemon shuts down partway through the session
+            # and another docker command is issued, we'll catch that here.
+            self.handle_docker_start_failure(exc_val)
+        # Use isinstance to catch subclasses of docker.errors.BuildError
+        elif isinstance(exc_val, docker.errors.BuildError):
+            self.handle_docker_build_failure(exc_val)
+
+    def handle_docker_build_failure(self, e: docker.errors.BuildError):
+        # If the user is in verbose mode, the build log has already been printed.
+        if not self.verbose:
+            for line in e.build_log:
+                typer.echo(line)
+
+        print_err(f"Fatal Docker build error: {e}\n" "Above is the full build log.\n")
+
+        if type(e) is DockerPreBuildFailure:
+            print_err(
+                "Garden could not set up your base Docker image. "
+                "If you supplied a requirements file, check that it's formatted correctly.\n"
+            )
+        elif type(e) is DockerBuildFailure:
+            last_line = e.build_log[-1] if len(e.build_log) > 0 else None
+            if "Traceback" in last_line:
```

Comment:
I think this `in last_line` check will break if it really is None

Reply:
Good catch, addressed